src/conversions.rs

```diff
@@ -30,77 +34,190 @@ impl Optimizer {
         }
     }
 
-    pub(crate) fn body_to_code(&mut self, expr: &Expr) -> Vec<Code> {
+    pub(crate) fn expr_to_structured_block(&mut self, expr: &Expr) -> StructuredBlock {
+        if let Expr::Call(func, args) = expr {
+            match (func.as_str(), &args.as_slice()) {
+                ("Block", [block]) => {
+                    StructuredBlock::Block(Box::new(self.expr_to_structured_block(block)))
+                }
+                ("Basic", [basic_block]) => {
+                    StructuredBlock::Basic(Box::new(self.expr_to_basic_block(basic_block)))
+                }
+                ("Ite", [name, then_branch, else_branch]) => StructuredBlock::Ite(
+                    name.to_string(),
+                    Box::new(self.expr_to_structured_block(then_branch)),
+                    Box::new(self.expr_to_structured_block(else_branch)),
+                ),
+                ("Loop", [block]) => {
+                    StructuredBlock::Loop(Box::new(self.expr_to_structured_block(block)))
+                }
+                ("Sequence", [block, rest]) => StructuredBlock::Sequence(vec![
+                    self.expr_to_structured_block(block),
+                    self.expr_to_structured_block(rest),
+                ]),
+                ("Break", [n]) => {
+                    if let Expr::Lit(egglog::ast::Literal::Int(n)) = n {
+                        StructuredBlock::Break(*n)
+                    } else {
+                        panic!("expected int literal for break");
```

Comment:
here, and elsewhere: might be helpful to add more info for debugging, like "expected __, got __"

Reply:
I usually add more info as I go (hopefully we never see this)